pkg/controller/customrule/customrule_controller.go

```diff
@@ -0,0 +1,212 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customrule
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ComplianceAsCode/compliance-operator/pkg/apis/compliance/v1alpha1"
+	"github.com/ComplianceAsCode/compliance-sdk/pkg/scanner"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// CustomRuleReconciler reconciles a CustomRule object
+type CustomRuleReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=compliance.openshift.io,resources=customrules,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=compliance.openshift.io,resources=customrules/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=compliance.openshift.io,resources=customrules/finalizers,verbs=update
+
+// Reconcile is the main reconciliation loop for CustomRule resources
+func (r *CustomRuleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling CustomRule")
+
+	// Fetch the CustomRule instance
+	rule := &v1alpha1.CustomRule{}
+	err := r.Get(ctx, req.NamespacedName, rule)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Info("CustomRule resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to get CustomRule")
+		return ctrl.Result{}, err
+	}
+
+	// Check if the rule has already been validated for the current generation
+	if rule.Status.ObservedGeneration == rule.Generation && rule.Status.Phase == v1alpha1.CustomRulePhaseReady {
+		log.Info("CustomRule already validated for current generation", "generation", rule.Generation)
+		return ctrl.Result{}, nil
+	}
+
+	// Perform validation
+	var validationErr error
+
+	// 1. Validate basic structure
+	if err := r.validateStructure(rule); err != nil {
+		validationErr = fmt.Errorf("structure validation failed: %w", err)
+	}
+
+	// 2. Validate and compile CEL expression using SDK's RuleBuilder
+	if validationErr == nil {
+		if err := r.validateCELExpressionWithBuilder(rule); err != nil {
+			validationErr = err
+		}
+	}
```

Comment:
Another good thing to potentially add here would be to validate the rule inputs, and see if the service account we're using has the ability to access those resources. If not, then we fail as early as possible so that users can adjust the rules before they run them in a scan.

Reply:
Perhaps compliance-sdk exposes an API like:

```go
  err := scanner.ValidatePermissionsForInputs(rule.Spec.Inputs)
```